README.md
### This project has been transferred to [DWTechs](https://github.com/DWTechs/CSVx.js). 

### Please update your project dependencies to the latest version [on the new repository](https://github.com/DWTechs/CSVx.js).


[![License: MIT](https://img.shields.io/npm/l/@lcluber/csvxjs.svg?color=brightgreen)](https://opensource.org/licenses/MIT)
[![npm version](https://badge.fury.io/js/%40lcluber%2Fcsvxjs.svg)](https://www.npmjs.com/package/@lcluber/csvxjs)
[![last version release date](https://img.shields.io/github/release-date/LCluber/CSVx.js)](https://www.npmjs.com/package/@lcluber/csvxjs)
![Jest:coverage](https://img.shields.io/badge/Jest:coverage-100%25-brightgreen.svg)
[![minified size](https://img.shields.io/bundlephobia/min/@lcluber/csvxjs?color=brightgreen)](https://www.npmjs.com/package/@lcluber/csvxjs)


## Synopsis

**[CSVx.js](https://github.com/LCluber/CSVx.js)** is an open source CSV library written in TypeScript.

## Motivation

The main purpose of this library is to provide an easy way to export your data as CSV or transform your CSV data to something else.

## Installation

### npm

```bash
$ npm i @lcluber/csvxjs
```

### yarn

```bash
$ yarn add @lcluber/csvxjs
```

## Usage

### ES6

```html
<button id="csv">Export CSV</button>
<div id="table"></div>
```

```javascript
import { Export, Convert } from '@lcluber/csvxjs';

// Convert an array to CSV file
let array = [
  {
    firstname:'Galileo',
    lastname:'Galiléi',
    born:1564,
    died:1642
  },
  {
    firstname:'Nikola',
    lastname:'Tesla',
    city:'Smiljan',
    born:1856,
    died:1943
  },
  {
    firstname:'Albert',
    born:1879,
    lastname:'Einstein',
    died:1955
  }
];
let customLabels = {
  firstname: 'First name',
  lastname: 'Last name', 
  city: 'City',
  born: 'Born',
  died: 'Died'
};
let exportButton = document.getElementById('csv');
exportButton.addEventListener('click', function() {
  Export.data('scientists',array,{separator: ';', customLabels: customLabels});// ; separator for excel friendly imports
});

// Convert CSV data to HTML table
let data = '"Firstname";"Lastname";"Born";"Died"\r\n\
"Galileo";"Galilei";"1564";"1642"\r\n\
"Nikola";"Tesla";"1856";"1943"\r\n\
"Albert";"Einstein";"1879";"1955"';

document.getElementById("table").innerHTML = Convert.table(data,{separator: ';'}, {table: 'table table-striped'});
```

### IIFE

```html
<script src="node-modules/@lcluber/csvxjs/dist/csvx.iife.min.js"></script>
<button id="csv">Export CSV</button>
<div id="table"></div>
```

```javascript

// Convert an array to CSV file
var array = [
 {
    firstname:'Galileo',
    lastname:'Galiléi',
    born:1564,
    died:1642
  },
  {
    firstname:'Nikola',
    lastname:'Tesla',
    city:'Smiljan',
    born:1856,
    died:1943
  },
  {
    firstname:'Albert',
    born:1879,
    lastname:'Einstein',
    died:1955
  }
];
var customLabels = {
  firstname: 'First name',
  lastname: 'Last name', 
  city: 'City',
  born: 'Born',
  died: 'Died'
};
var exportButton = document.getElementById('csv');
exportButton.addEventListener('click', function() {
  CSVx.Export.data('scientists',array,{separator: ';', customLabels: customLabels});// ; separator for excel friendly imports
});

// Convert CSV data to HTML table
var data = '"Firstname";"Lastname";"Born";"Died"\r\n\
"Galileo";"Galilei";"1564";"1642"\r\n\
"Nikola";"Tesla";"1856";"1943"\r\n\
"Albert";"Einstein";"1879";"1955"';

document.getElementById("table").innerHTML = CSVx.Convert.table(data,{separator: ';'}, {table: 'table table-striped'});
```

## API Reference

```javascript

class Convert {
  static setOptions(options: Partial<Options>): void;
  static setCSS(css: Partial<CSS>): void;
  // convert CSV to Javascript array
  static array(data: string, options?: Partial<Options>, css?: Partial<CSS>): Array<Array<string>> | false;
  // convert CSV to HTML table
  static table(data: string, options?: Options, css?: CSS): string | false;
}

class Export {
  // Exort CSV file
  static data(filename: string, data: Data[], options?: Partial<Options>): boolean;
  static setOptions(options: Partial<Options>): void;
}

interface Data { [key: string]: number|string }[]

interface Options {
  data?: string; // default : 'text/csv'
  charset?: string; // default : 'utf-8'
  labels?: boolean; // default : true
  quote?: string; // default : '"'
  separator?: string; // default : ','
  CRLF?: string; // default : '\r\n'
  customLabels: { [key: string]: string }; // default : {}
}

interface CSS {
  table?: string; // default : ''
  th?: string; // default : ''
}

```

## Contributors

CSVx.js is still in early development and I would be glad to get all the help you can provide for this project.
To contribute you can clone the project on **[GitHub](https://github.com/LCluber/CSVx.js)** and See **NOTICE.md** for detailed installation walkthrough.

## License

MIT License

Copyright (c) 2018 Ludovic CLUBER

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
Version 0.2.3 (February 18th 2021)
-----------------------------
- Project transferred from LCluber to DWTechs account

Version 0.2.2 (February 17th 2021)
-----------------------------
 * Updated Ch.js to Checkhard.js dependency

Version 0.2.1 (December 12th 2020)
-----------------------------
 * Fix declaration file
 * Lighter library 
 * Delete Mouette.js and Wee.js devDependencies

Version 0.2.0 (November 17th 2020)
-----------------------------
 * Add support for array of different objects 

Version 0.1.11 (April 10th 2020)
-----------------------------
 * Fix csv export with custom labels on Firefox

Version 0.1.10 (April 10th 2019)
-----------------------------
 * Fix csv export with better filename checking

Version 0.1.9 (April 09th 2019)
-----------------------------
 * Fix csv export with special characters

Version 0.1.8 (March 31th 2019)
-----------------------------
 * Improve typings

Version 0.1.7 (March 13th 2019)
-----------------------------
 * Add byte order mark U+FEFF at the beginning of the data stream

Version 0.1.6 (March 01st 2019)
-----------------------------
 * Default name to 'export.csv' if filename parameter is empty

Version 0.1.5 (February 11th 2019)
-----------------------------
 * Add custom labels option
 * array() method of Convert class deletes empty strings

Version 0.1.4 (February 08th 2019)
-----------------------------
 * Add Convert class to transform CSV data to array or HTML table

Version 0.1.3 (December 23th 2018)
-----------------------------
 * Add Ch.js dependency
 * Update Mouette.js dependency to version 0.2.8
 * Update Wee.js dependency to version 0.2.1
 * Update README.md

Version 0.1.2 (December 07th 2018)
-----------------------------
 * Added logs with @lcluber/Mouettejs

Version 0.1.1 (December 2nd 2018)
-----------------------------
 * fromObject method now called data

Version 0.1.0 (December 1st 2018)
-----------------------------
 * Initial version


web/js/example.js
var customLabels = ['First name', 'Last name', 'city', 'Born', 'Died'];
var array = [
  {
    firstname:'Galileo',
    lastname:'Galiléi',
    born:1564,
    died:1642
  },
  {
    firstname:'Nikola',
    lastname:'Tesla',
    city:'Smiljan',
    born:1856,
    died:1943
  },
  {
    firstname:'Albert',
    born:1879,
    lastname:'Einstein',
    died:1955
  }
];
var customLabels = {
  firstname: 'First name',
  lastname: 'Last name', 
  city: 'City',
  born: 'Born',
  died: 'Died'
};
var exportButton = Wee.Dom.findById('csv');
exportButton.addEventListener('click', function() {
  CSVx.Export.data('scientists',array, {separator:';', customLabels: customLabels });
});

var data = '"Firstname";"Lastname";"Born";"Died"\r\n\
"Galileo";"Galilei";"1564";"1642"\r\n\
"Nikola";"Tesla";"1856";"1943"\r\n\
"Albert";"Einstein";"1879";"1955"\r\n\ ';
 document.getElementById("table").innerHTML = CSVx.Convert.table(data,{separator: ';'}, {table: 'table table-striped'});